File: MAL_HELPER.py
# !/usr/bin/env python3
# -*- coding: utf-8 -*-

import requests
import time
from collections import deque
from datetime import datetime
from lxml import etree, html
from urllib.request import urlopen


def write_log(f_file, f_text):

    with open(f_file, "ab") as f:
        f.write(f_text.encode("utf8"))
        f.write("\n".encode("utf8"))


def get_list(f_list, f_user, f_site, f_file):

    for item_type in (
            "anime",
            "manga",
    ):

        link = "{0}/malappinfo.php?u={1}&status=all&type={2}".format(
            f_site,
            f_user,
            item_type,
        )
        tree = etree.parse(urlopen(link))
        root_xml = tree.getroot()
        log_text = "{0}\tparsing of xml from url \"{1}\".".format(
            datetime.now().time(),
            link,
        )
        print(log_text,)

        for item in root_xml.iter("series_{0}db_id".format(item_type,)):

            try:
                item_id = int(item.text)

            except ValueError:
                item_id = item.text
                log_text = "{0}\t{1}_id={2} is not int.\tValueError".format(
                    datetime.now().time(),
                    item_type,
                    item_id,
                )
                write_log(f_file, log_text)

            f_list.append((item_type, item_id,))
            log_text = "{0}\tadded {1}_id={2} to list.".format(
                datetime.now().time(),
                item_type,
                item_id,
            )
            print(log_text,)

        log_text = "{0}\tparsing of {1} list is complete.".format(
            datetime.now().time(),
            item_type,
        )
        print(log_text,)

    log_text = "{0}\tgetting list is complete.".format(
        datetime.now().time(),
    )
    print(log_text,)

    return f_list


def get_page(f_site, f_task, f_file):

    link = "{0}/{1}/{2}/".format(
        f_site,
        f_task[0],
        f_task[1],
    )
    page = requests.get(
        link
    )

    # error 404 (not found)
    if page.status_code == 404:
        log_text = "{0}\tpage \"{1}\" not found.\tpage not found".format(
            datetime.now().time(),
            link,
        )
        print(
            log_text,
        )
        write_log(
            f_file,
            log_text
        )
        return None

    # error 429 (too many requests)
    counter = 0
    while page.status_code == 429:

        if counter == 9:
            log_text = "{0}\tpage \"{1}\" not returned 10 times.\tpage not returned".format(
                datetime.now().time(),
                link,
            )
            write_log(
                f_file,
                log_text
            )
            break

        counter += 1
        log_text = "{0}\tpage \"{1}\" not returned. Repeat GET request in 5 seconds.".format(
            datetime.now().time(),
            link,
        )
        print(
            log_text,
        )
        time.sleep(
            5
        )
        page = requests.get(
            link
        )

    return page


def parse_page(
        f_page,
        f_file
):

    root_html = html.fromstring(
        f_page.content
    )
    f_set = set()

    for attr in root_html.xpath(
            '//table[@class="anime_detail_related_anime"]//a/@href'
    ):

        f_list = attr.split(
            '/'
        )
        item_type = f_list[1]

        try:
            item_id = int(
                f_list[2]
            )

        except ValueError:

            if f_list[2] == "":
                continue

            item_id = f_list[2]
            log_text = "{0}\t{1}_id={2} is not int.\tValueError".format(
                datetime.now().time(),
                item_type,
                item_id,
            )
            write_log(
                f_file,
                log_text
            )

        task = (
            item_type,
            item_id,
        )
        f_set.add(
            task
        )

    return f_set


def main(user):

    site = "https://myanimelist.net"

    user_list = list()
    visited_set = set()
    queue = deque()

    file_result = "result.tsv"
    file_error_log = "errors.tsv"

    result_file = open(
        file_result,
        "wb"
    )
    result_file.close()
    error_log_file = open(
        file_error_log,
        "wb"
    )
    error_log_file.close()

    user_list.extend(
        get_list(
            user_list,
            user,
            site,
            file_error_log
        )
    )
    queue.extend(
        user_list
    )

    while queue:

        task = queue.popleft()

        if task in visited_set:
            continue

        if task not in user_list:
            log_text = "{0}\t{1} added to result_set.".format(
                datetime.now().time(),
                task,
            )
            print(
                log_text
            )
            log_text = "{0}\t{1}\t{2}/{0}/{1}/".format(
                task[0],
                task[1],
                site,
            )
            write_log(
                file_result,
                log_text
            )

        visited_set.add(task)
        log_text = "{0}\t{1} added to visited_set.".format(
            datetime.now().time(),
            task,
        )
        print(
            log_text,
        )

        page = get_page(
            site,
            task,
            file_error_log
        )

        if page is None:
            continue

        else:
            urls_set = parse_page(
                page,
                file_error_log
            )

        for item in urls_set:

            if item not in visited_set:
                queue.append(
                    item
                )
                log_text = "{0}\t{1} added to queue.".format(
                    datetime.now().time(),
                    item,
                )
                print(
                    log_text
                )

    log_text = "{0}\tqueue is empty, mission complete.".format(
        datetime.now().time()
    )
    print(log_text)

    return None

if __name__ == "__main__":
    main("XuT")

File: mal_helper.java
import java.io.File;
import java.io.IOException;
import java.io.FileWriter;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.PriorityQueue;
import java.util.Date;
import org.jsoup.Connection;
import org.jsoup.Connection.Response;
import org.jsoup.Connection.Method;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

public class searchRelatedLinks {

    public static void main(String[] args) throws IOException {

        String site = "https://myanimelist.net";
        String user = "user";
        String[] list_type = {"anime", "manga"};

        SimpleDateFormat time_format_print = new SimpleDateFormat("HH:mm:ss.SSSSS");
        SimpleDateFormat time_format_file = new SimpleDateFormat("HH-mm-ss");

        ArrayList<String> user_list = new ArrayList<>();
        HashSet<String> visited_set = new HashSet<>();
        PriorityQueue<String> queue = new PriorityQueue<>();

        String log;
        String link;
        Response page;
        try {
            File file_error_log = new File("./".concat(getTime(time_format_file)).concat("_").concat("errors.tsv"));
            File file_result = new File("./".concat(getTime(time_format_file)).concat("_").concat("result.tsv"));
            if (file_error_log.delete() && file_result.delete()) {
                log = getTime(time_format_print).concat("\tFiles deleted.");
                System.out.println(log);
            } else if (file_error_log.createNewFile() && file_result.createNewFile()) {
                log =getTime(time_format_print).concat("\tFiles created.");
                System.out.println(log);
            }
            FileWriter error_writer = new FileWriter(file_error_log, true);
            FileWriter result_writer = new FileWriter(file_result, true);

            for (String type : list_type) {
                link = site.concat("/malappinfo.php?u=").concat(user).concat("&status=all&type=").concat(type);
                page = getPage(link, error_writer, time_format_print);
                log = getTime(time_format_print).concat("\tTry to parse page \"").concat(link).concat("\".");
                System.out.println(log);
                ArrayList<String> temp_list = parsePage(page, type, time_format_print);
                user_list.addAll(temp_list);
                log = getTime(time_format_print).concat("\t").concat(Integer.toString(temp_list.size())).concat(" tasks added to queue.");
                System.out.println(log);
            }
            queue.addAll(user_list);
            String task;
            while ((task = queue.poll()) != null) {
                if (visited_set.contains(task)) {
                    log = getTime(time_format_print).concat("\tVisited_set already contains task \"").concat(task).concat("\". Go to next loop.");
                    System.out.println(log);
                    continue;
                } else {
                    visited_set.add(task);
                    log = getTime(time_format_print).concat("\tTask \"").concat(task).concat("\" added to visited_set.");
                    System.out.println(log);
                }
                String[] title = task.split("/");
                String title_type = title[0];
                String title_id = title[1];
                if (!user_list.contains(task)) {
                    log = title_type.concat("\t").concat(title_id).concat("\t").concat(site).concat("/").concat(task).concat("/");
                    writeFile(result_writer, log);
                    log = getTime(time_format_print).concat("\tTask \"").concat(task).concat("\" wrote to result_file.");
                    System.out.println(log);
                }
                link = site.concat("/").concat(task).concat("/");
                page = getPage(link, error_writer, time_format_print);
                if (page != null) {
                    log = getTime(time_format_print).concat("\tTry to parse page \"").concat(link).concat("\".");
                    System.out.println(log);
                    ArrayList<String> urls = parsePage(page, "title", time_format_print);
                    for (String item : urls) {
                        if (!visited_set.contains(item)) {
                            queue.add(item);
                            log = getTime(time_format_print).concat("\tTask \"").concat(item).concat("\" added to queue.");
                            System.out.println(log);
                        } else {
                            log = getTime(time_format_print).concat("\tVisited_set already contains task \"").concat(item).concat("\". Go to next loop.");
                            System.out.println(log);
                        }
                    }
                }

            }
            log = getTime(time_format_print).concat("\tQueue is empty, mission complete.");
            System.out.println(log);
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private static String getTime(
            SimpleDateFormat time_format
    ) {
        Date date = new Date();
        return time_format.format(date);
    }

    private static void writeFile(
        FileWriter file,
        String text
    ) {
        try {
            file.write(text);
            file.append('\n');
            file.flush();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static Response getPage(
            String link,
            FileWriter error_writer,
            SimpleDateFormat time_format
    ) {
        Response response = null;
        String log;
        try {
            log = getTime(time_format).concat("\tTry to get page \"").concat(link).concat("\".");
            System.out.println(log);
            Connection connection = Jsoup.connect(link).ignoreHttpErrors(true).userAgent("Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36").maxBodySize(0).timeout(0);
            response = connection.method(Method.GET).execute();
            while (response.statusCode() != 200) {
                if (response.statusCode() == 404) {
                    log = getTime(time_format).concat("\tpage \"").concat(link).concat("\" not found.\tpage not found");
                    System.out.println(log);
                    writeFile(error_writer, log);
                    break;
                }
                connection = Jsoup.connect(link).ignoreHttpErrors(true).userAgent("Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.87 Safari/537.36").maxBodySize(0).timeout(0);
                response = connection.method(Method.GET).execute();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (response == null) {
            log = getTime(time_format).concat("\tGet page \"").concat(link).concat("\" failed.");
            System.out.println(log);
        } else {
            log = getTime(time_format).concat("\tGet page \"").concat(link).concat("\" succeed.");
            System.out.println(log);
        }
        return response;
    }

    private static ArrayList<String> parsePage(
            Response response,
            String parse_type,
            SimpleDateFormat time_format
    ) {
        ArrayList<String> list = new ArrayList<>();
        try {
            Document document = response.parse();
            if (parse_type.equals("anime") || parse_type.equals("manga")) {
                Elements elements = document.getElementsByTag("series_".concat(parse_type).concat("db_id"));
                for (Element element : elements) {
                    list.add(parse_type.concat("/").concat(element.text()));
                }
            }
            if (parse_type.equals("title")) {
                Elements elements = document.select("table.anime_detail_related_anime a");
                for (Element element : elements) {
                    String[] link = element.attr("href").split("/", 4);
                    String item_type = link[1];
                    String item_id = link[2];
                    if (item_id.equals("")) {
                        continue;
                    }
                    if (item_type.equals("anime") || item_type.equals("manga")) {
                        list.add(item_type.concat("/").concat(item_id));
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        String log = getTime(time_format).concat("\tParsing complete.");
        System.out.println(log);
        return list;
    }

}
